docs/proposals/201807_store_instance_high_availability.md

```diff
@@ -0,0 +1,131 @@
+sidecar servers to return their data (subject to timeouts) before returning a
+response to a client.
+
+When retrieving data from Thanos bucket store instances, however, the desired
+behaviour is different; we want Thanos use the first successful response it
+receives, on the assumption that all bucket store instances that communicate
+with the same bucket have access to the same data.
+
+To support the desired behaviour for bucket store instances while still
+allowing for deduplication, we propose to expand the [InfoResponse
+Protobuf](https://github.com/improbable-eng/thanos/blob/b67aa3a709062be97215045f7488df67a9af2c66/pkg/store/storepb/rpc.proto#L28-L32)
+used by the Store API by adding a single boolean field to indicate whether the
+store instance in question is acting as a 'gateway':
+
+```diff
+--- before	2018-07-02 15:49:09.000000000 +0100
++++ after	2018-07-02 15:49:13.000000000 +0100
+@@ -1,5 +1,6 @@
+ message InfoResponse {
+   repeated Label labels = 1 [(gogoproto.nullable) = false];
+   int64 min_time        = 2;
+   int64 max_time        = 3;
++  bool gateway          = 4;
+ }
+```
+
+Thanos bucket store instances (i.e. store instances that act as 'gateways' to
+AWS S3 or Google Cloud Storage) will set `gateway` to `true`. A `bool` type in
+Protobuf defaults to false, so the behaviour of other existing store instances
+that do not explicitly set a value for `gateway` will not be affected.
+
+If a store instance is a gateway, query instances will treat each store
+instance in a label group as having access to the same data. Query instances
+will randomly pick any two store instances[1][] from the same gateway group and
+use the first response returned.
+
+[1]: https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf
+
+Otherwise, query instances will wait for all replicas within the same
+label group to respond (subject to existing timeouts) before returning a
+response, consistent with the current behaviour.
+
+### Scope
+
+Horizontal scaling should be handled separately and is out of scope for this proposal.
+
+## User experience
+
+From a user's point of view, query responses should be faster and more reliable:
+
+- Running multiple bucket store instances will allow the query to be served even
+  if a single store instance fails.
+
+- Query latency should be lower since the response will be served from the
+  first bucket store instance to reply.
+
+The user experience for query responses involving only Thanos sidecars will be
+unaffected.
+
+## Alternatives considered
+
+### Implicitly relying on store labels
+
+Rather than expanding the `InfoResponse` Protobuf, we had originally considered relying on
+an empty set of store labels to determine that a store instance was acting as a gateway.
+
+We decided against this approach as it would make debugging harder due to its
+implicit nature, and is likely to cause bugs in future.
+
+## Open issues
+
+### Querying from multiple buckets or object stores
```

Comment:
I think I personally prefer this approach, would allow us do run queries that span multiple buckets.

Example - we collect info from two tenants each of which gets uploaded to their own bucket `tenantA` and `tenantB` if we wanted to query the metrics for both adding the a `Identifier` to the `InfoResponse` we could have the bucket name on there and it would be valid to wait for both responses.

A further use case of this could be to have buckets that are historical / archive data and some that are within x days and would mean that you could spin up store nodes quicker for the recent bucket as would require less overhead at start.

Reply:
Agree here with @domgreen. Maybe slight change in `gateway: bool` into something like `gateway string` with bucket name would allow us for basic functional sharding of stores in addition HA. What do you think @mattbostock ?

Reply:
it makes sense to having a string instead of bool

Reply:
I agree, I think using a string identifier is the way forward.

The Protobuf definition could be:

```
bool deduplicated = 4
string store_group_id = 5
```

The `deduplicated` boolean would determine how the query instance handles retrieving the data (i.e. whether to wait for all results). I think I prefer `deduplicated` over `gateway` as it's more specific to the functionality we care about in this case. Sidecar instances would have `deduplicated` set to false. Bucket stores would have `deduplicated` set to true.

It'd be up to the specific implementation to determine what the `store_group_id` string is. For bucket stores, I think concatenating the object store URL and bucket name would make sense. For all other stores, I'd leave it empty for now until we have a good reason to use it.

The logic would work like this:

```
if deduplicated:
  query_and_use_fastest_response_per_store_group
else
   query_and_wait_for_all_responses
```

Stores would be grouped by `store_group_id`. Stores having an empty `store_group_id` field would be considered all part of the same group.

Since boolean fields are false by default in Protobufs, `deduplicated` would be false and existing instances would be treated as before for backwards compatibility (i.e. the query instance will wait for all responses, subject to timeouts).

If you agree, I'll update the proposal to match.

Reply:
do we really need a redundant `deduplicated` option? Isn't it enough having specified `store_group_id` as a object store+bucket_name as identifier? Thus, we don't have to wait for all responses if stores have the same `store_group_id`.

Reply:
I think a boolean field is necessary to avoid implicit behaviour when `store_group_id` is an empty string. Protobuf v3 doesn't allow null values if I've understood correctly, otherwise we could use have used a null value for `store_group_id` (but that's still implicit behaviour).

I'd welcome suggestions for a better name than `deduplicated` - I'm aiming for a boolean that we can default to false for backwards compatibility with existing peers in the cluster running older versions.

Reply:
we can account only non-zero string in `store_group_id` for the deduplication, hence older version would work as before

Reply:
I think relying on an empty string is too implicit as it alters the behaviour of how responses are treated.

Reply:
I don't think this's a disadvantage, if you describe the behaviour in the documentation and the help. Otherwise you have to bear in mind both options to be working as expected.

Reply:
Sorry for the massive delay in reviews and thanks for this (: 

`store_group_id` makes sense to me. `deduplicated` is tricky. I don't like it because both store gateways with `query_and_use_fastest_response_per_store_group` and sidecars with `query_and_wait_for_all_responses` logic are kind of duplicated, so we can work on naming definitely. 

However, what about another idea: 
In `Info` let's add:
- `store_group_id` 
- `peer_type` (like it is in gossip `cluster.PeerState.Type`). 

I am proposing this because I am trying to figure out how to make deduplication flow consistent across ALL components. Including not yet solved: `Ruler` deduplication! 

Basically, with this we can deduct correct behavior everywhere:
* Store gateway: `store_group_id` would be as @mattbostock suggested and peerType=store so query can perform `query_and_use_fastest_response_per_store_group` logic.
* Sidecars: `store_group_id` would be all external labels as it is ID of those. peerType=source so query can assume `query_and_wait_for_all_responses` flow with some `replica` label deduplication.
* Ruler: The tricky part is that we might want these in HA, but it's hard to define external labels for these, so having some arbitrary `store_group_id` would be perfect. Plus ruler is `peerType=source` so same replica deduplication might apply here as well.

I am not saying we want solve ruler with this proposal, just that we might want to reuse same thing and move ALL to `store_group_id` for deduplication logic for consistency. 

Tricky part here:
4 Rules (in 2 different groups) - it would be not trivial to implement this, because the `query_and_wait_for_all_responses` + `dedup` is done truly by removing (or not) replica label. But that is something not for proposal anyway.

What do you think?

Reply:
Thanks for the comments all.

@Bplotka: I think your proposal makes sense, I'll update this PR to use that.